include/libtorrent/hasher512.hpp

```diff
@@ -0,0 +1,124 @@
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in
+      the documentation and/or other materials provided with the distribution.
+    * Neither the name of the author nor the names of its
+      contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_HASHER512_HPP_INCLUDED
+#define TORRENT_HASHER512_HPP_INCLUDED
+
+#include "libtorrent/config.hpp"
+#include "libtorrent/sha1_hash.hpp"
+#include "libtorrent/span.hpp"
+
+#include <cstdint>
+
+#ifdef TORRENT_USE_LIBGCRYPT
+#include <gcrypt.h>
+
+#elif TORRENT_USE_COMMONCRYPTO
+#include <CommonCrypto/CommonDigest.h>
+
+#elif TORRENT_USE_CRYPTOAPI
+#include <windows.h>
+#include <wincrypt.h>
+
+#elif defined TORRENT_USE_LIBCRYPTO
+
+extern "C" {
+#include <openssl/sha.h>
+}
+
+#else
+#include "libtorrent/sha512.hpp"
+#endif
+
+namespace libtorrent
+{
+	using sha512_hash = digest32<512>;
+
+	// this is a SHA-512 hash class.
+	//
+	// You use it by first instantiating it, then call ``update()`` to feed it
+	// with data. i.e. you don't have to keep the entire buffer of which you want to
+	// create the hash in memory. You can feed the hasher parts of it at a time. When
+	// You have fed the hasher with all the data, you call ``final()`` and it
+	// will return the sha1-hash of the data.
+	//
+	// The constructor that takes a ``char const*`` and an integer will construct the
+	// sha1 context and feed it the data passed in.
+	//
+	// If you want to reuse the hasher object once you have created a hash, you have to
+	// call ``reset()`` to reinitialize it.
+	//
+	// The built-in software version of the sha512-algorithm is from LibTomCrypt
+	// For more info, see ``src/sha512.cpp``.
+	class TORRENT_EXPORT hasher512
```

Comment:
@aldenml is there a good reason for this being public? afaict, it's only used in the ed25519 implementation. I wonder if it would make sense to make this private and move it into the `ed25519` source directory (along with `sha512.cpp`/`sha512.hpp`)

Reply:
I remember that I left the `TORRENT_EXPORT` (and left it public as `hasher`) because I wanted to have them available outside `libtorrent`, they are very convenient to use, but it turn out I never actually use them. So I guess there is no a good reason, and maybe `hasher` should be private too.